flat_mutation_reader.hh

```diff
@@ -200,7 +205,7 @@ public:
                     next_partition();
                     continue;
                 }
-                if (consumer(std::move(mf)) == stop_iteration::yes) {
+                if (filter(mf) && (consumer(std::move(mf)) == stop_iteration::yes)) {
```

Comment:
This is not a filter, but a stop condition.

Reply:
Hmm, the stop condition was there before.  With the filter in the expression, `consumer(std::move(mf))` will not be called if the filter returns `false` and the mutation fragment will not be consumed at all.
That said, it is true though that at this time we have no use for this method as a conditional filter and it could return void.

Reply:
More than that, stopping in the middle of a partition is not legal, so we cannot stop here on any fragment.

Reply:
@denesb The consumer can stop after `partition_end`, no?

Reply:
Yes, *only* after `partition_end`.

BTW my top comment here refers to the fact that this filter is now is in very not filter-y roles. It now determines when consuming stops, as opposed to allowing/dropping fragments or partitions.

Reply:
Previously the `consumer` could return `stop_iteration::yes` on any `mutation_fragment`.  Doing so after mutation_fragments that aren't `partition_end` is a bug indeed but it was possible before, and will be detected in `~mutation_fragment_stream_validator()`.
Now,  stopping the stream would require both returning `true` by the filter and `stop_iteration::yes` from the `consumer`.  If the mutation filter returns `false`, the mutation_fragment will be skipped (dangerous indeed) and the stream will not stop.

The mutation_fragment can't be both skipped and consumed (for stopping consumption), that doesn't make sense at all.  As I said, we can have the mutation filter return `void` rather than `bool` and consume all mutation_fragments with no change with that regards..

Reply:
I just realized this is `consume_pausable()`, where you can stop anywhere you want, the assumption is that you are not stopping for good, just "pausing". Calling this method again, you will continue where you left off previously.

So when the filter rejects a fragment, it just throws out the consumer from the consumption loop. The consumer then might call `consume_pausable()` again, essentially just skipping a fragment. Or it might consider the consumption to be over. In both cases it can end up with an invalid fragment stream without even knowing it.

I don't know what to think about this. On one hand its great that we are reusing the existing filter and consume infrastructure for validation. On the other hand we are just about to build into it a huge gun aimed at our feet.

Reply:
@denesb When the filter rejects a fragment the consumption loop is not terminated, it continues. It's equivalent to consuming a filtered stream:

`(read() | filter()).consume_pausable_in_thread()`

I don't understand what's the objection for this.

Reply:
It is terminated:
```
if (filter(mf) && (consumer(std::move(mf)) == stop_iteration::yes))
```

For the consumption loop to continue the filter has to return true and the consumer has to opt to continue. If the filter rejects the fragment the consumer will not even see it and the consumption loop will terminate.

Reply:
I was wrong, its the other way around. Both have to opt in for the consumption to terminate. I guess that is fine then.

Reply:
Wait, shouldn't this be
```
if (!filter(mf) && (consumer(std::move(mf)) == stop_iteration::yes))
```
then? `true` means the fragment passed the filter, no? In the current form the consumption will break when the fragment passes the filter *and* the consumer wants to stop.

Reply:
Right. If the filter rejects the fragment, then it won't be passed to the consumer and the loop will continue. What's wrong with that?

Reply:
I think Benny's code is correct, if `!filter(mf)` then the fragment is treated as if it didn't exist, so we don't call the consumer, and just keep looping. If `filter(mf) == true`, we execute whatever logic we had before for that fragment, cause it exists in the consumer's input stream.

Reply:
Ok. Fair enough.